# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditHelpers.h
# Commit: a43e95a534eb
# Full Hash: a43e95a534eb5c13883cc2a61fb64c15fe29e674
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-10-05 03:40:21
# Regressor Bug: 1792654
# File Overlap Count: 1
# Description:
#   Bug 1792654 - Make `CreateNodeResultBase`, `MoveNodeResult`, `SplitNodeResult` and `SplitRangeOffResult` handle caret point with a common base class r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D158335
# ==============================================================================

diff -r 18196c5fab62 -r a43e95a534eb editor/libeditor/HTMLEditHelpers.h
--- a/editor/libeditor/HTMLEditHelpers.h	Tue Oct 04 21:13:53 2022 +0000
+++ b/editor/libeditor/HTMLEditHelpers.h	Tue Oct 04 21:36:57 2022 +0000
@@ -17,7 +17,7 @@
 #include "mozilla/ContentIterator.h"
 #include "mozilla/EditorDOMPoint.h"
 #include "mozilla/EditorForwards.h"
-#include "mozilla/EditorUtils.h"  // for SuggestCaretOption(s)
+#include "mozilla/EditorUtils.h"  // for CaretPoint
 #include "mozilla/IntegerRange.h"
 #include "mozilla/Maybe.h"
 #include "mozilla/RangeBoundary.h"
@@ -76,7 +76,7 @@
  * This stores whether it's handled or not, and next insertion point and a
  * suggestion for new caret position.
  *****************************************************************************/
-class MOZ_STACK_CLASS MoveNodeResult final {
+class MOZ_STACK_CLASS MoveNodeResult final : public CaretPoint {
  public:
   constexpr bool Handled() const { return mHandled; }
   constexpr bool Ignored() const { return !Handled(); }
@@ -96,39 +96,6 @@
    */
   void MarkAsHandled() { mHandled = true; }
 
-  /**
-   * Suggest caret position to aHTMLEditor.
-   */
-  [[nodiscard]] MOZ_CAN_RUN_SCRIPT nsresult SuggestCaretPointTo(
-      const HTMLEditor& aHTMLEditor, const SuggestCaretOptions& aOptions) const;
-
-  /**
-   * IgnoreCaretPointSuggestion() should be called if the method does not want
-   * to use caret position recommended by this instance.
-   */
-  void IgnoreCaretPointSuggestion() const { mHandledCaretPoint = true; }
-
-  bool HasCaretPointSuggestion() const { return mCaretPoint.IsSet(); }
-  constexpr EditorDOMPoint&& UnwrapCaretPoint() {
-    mHandledCaretPoint = true;
-    return std::move(mCaretPoint);
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const SuggestCaretOptions& aOptions) {
-    MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
-    MOZ_ASSERT(
-        !aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt));
-    if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
-        !mCaretPoint.IsSet()) {
-      return false;
-    }
-    aPointToPutCaret = UnwrapCaretPoint();
-    return true;
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const HTMLEditor& aHTMLEditor,
-                        const SuggestCaretOptions& aOptions);
-
   MoveNodeResult(const MoveNodeResult& aOther) = delete;
   MoveNodeResult& operator=(const MoveNodeResult& aOther) = delete;
   MoveNodeResult(MoveNodeResult&& aOther) = default;
@@ -138,9 +105,9 @@
     MOZ_ASSERT(this != &aOther);
     // aOther is merged with this instance so that its caret suggestion
     // shouldn't be handled anymore.
-    aOther.mHandledCaretPoint = true;
+    aOther.IgnoreCaretPointSuggestion();
     // Should be handled again even if it's already handled
-    mHandledCaretPoint = false;
+    UnmarkAsHandledCaretPoint();
 
     mHandled |= aOther.mHandled;
 
@@ -148,15 +115,15 @@
     mNextInsertionPoint = aOther.mNextInsertionPoint;
 
     // Take the new caret point if and only if it's suggested.
-    if (aOther.mCaretPoint.IsSet()) {
-      mCaretPoint = aOther.mCaretPoint;
+    if (aOther.HasCaretPointSuggestion()) {
+      SetCaretPoint(aOther.CaretPointRef());
     }
     return *this;
   }
 
 #ifdef DEBUG
   ~MoveNodeResult() {
-    MOZ_ASSERT_IF(Handled(), !mCaretPoint.IsSet() || mHandledCaretPoint);
+    MOZ_ASSERT_IF(Handled(), !HasCaretPointSuggestion() || CaretPointHandled());
   }
 #endif
 
@@ -211,55 +178,54 @@
  private:
   explicit MoveNodeResult(const EditorDOMPoint& aNextInsertionPoint,
                           bool aHandled)
-      : mNextInsertionPoint(aNextInsertionPoint),
+      : CaretPoint(),
+        mNextInsertionPoint(aNextInsertionPoint),
         mHandled(aHandled && aNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
   explicit MoveNodeResult(EditorDOMPoint&& aNextInsertionPoint, bool aHandled)
-      : mNextInsertionPoint(std::move(aNextInsertionPoint)),
+      : CaretPoint(),
+        mNextInsertionPoint(std::move(aNextInsertionPoint)),
         mHandled(aHandled && mNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
   explicit MoveNodeResult(const EditorDOMPoint& aNextInsertionPoint,
                           const EditorDOMPoint& aPointToPutCaret)
-      : mNextInsertionPoint(aNextInsertionPoint),
-        mCaretPoint(aPointToPutCaret),
+      : CaretPoint(aPointToPutCaret),
+        mNextInsertionPoint(aNextInsertionPoint),
         mHandled(mNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
   explicit MoveNodeResult(EditorDOMPoint&& aNextInsertionPoint,
                           const EditorDOMPoint& aPointToPutCaret)
-      : mNextInsertionPoint(std::move(aNextInsertionPoint)),
-        mCaretPoint(aPointToPutCaret),
+      : CaretPoint(aPointToPutCaret),
+        mNextInsertionPoint(std::move(aNextInsertionPoint)),
         mHandled(mNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
   explicit MoveNodeResult(const EditorDOMPoint& aNextInsertionPoint,
                           EditorDOMPoint&& aPointToPutCaret)
-      : mNextInsertionPoint(aNextInsertionPoint),
-        mCaretPoint(std::move(aPointToPutCaret)),
+      : CaretPoint(std::move(aPointToPutCaret)),
+        mNextInsertionPoint(aNextInsertionPoint),
         mHandled(mNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
   explicit MoveNodeResult(EditorDOMPoint&& aNextInsertionPoint,
                           EditorDOMPoint&& aPointToPutCaret)
-      : mNextInsertionPoint(std::move(aNextInsertionPoint)),
-        mCaretPoint(std::move(aPointToPutCaret)),
+      : CaretPoint(std::move(aPointToPutCaret)),
+        mNextInsertionPoint(std::move(aNextInsertionPoint)),
         mHandled(mNextInsertionPoint.IsSet()) {
     AutoEditorDOMPointChildInvalidator computeOffsetAndForgetChild(
         mNextInsertionPoint);
   }
 
   EditorDOMPoint mNextInsertionPoint;
-  // Recommended caret point after moving a node.
-  EditorDOMPoint mCaretPoint;
   bool mHandled;
-  bool mutable mHandledCaretPoint = false;
 };
 
 /*****************************************************************************
@@ -267,7 +233,7 @@
  * HTMLEditor::SplitNodeDeepWithTransaction().
  * This makes the callers' code easier to read.
  *****************************************************************************/
-class MOZ_STACK_CLASS SplitNodeResult final {
+class MOZ_STACK_CLASS SplitNodeResult final : public CaretPoint {
  public:
   bool Handled() const { return mPreviousNode || mNextNode; }
 
@@ -384,55 +350,6 @@
     return EditorDOMPointType::After(mPreviousNode);
   }
 
-  /**
-   * Suggest caret position to aHTMLEditor.
-   */
-  [[nodiscard]] MOZ_CAN_RUN_SCRIPT nsresult SuggestCaretPointTo(
-      const HTMLEditor& aHTMLEditor, const SuggestCaretOptions& aOptions) const;
-
-  /**
-   * IgnoreCaretPointSuggestion() should be called if the method does not want
-   * to use caret position recommended by this instance.
-   */
-  void IgnoreCaretPointSuggestion() const { mHandledCaretPoint = true; }
-
-  bool HasCaretPointSuggestion() const { return mCaretPoint.IsSet(); }
-  constexpr EditorDOMPoint&& UnwrapCaretPoint() {
-    mHandledCaretPoint = true;
-    return std::move(mCaretPoint);
-  }
-  bool CopyCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const SuggestCaretOptions& aOptions) const {
-    MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
-    MOZ_ASSERT(
-        !aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt));
-    if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
-        !mCaretPoint.IsSet()) {
-      return false;
-    }
-    mHandledCaretPoint = true;
-    aPointToPutCaret = mCaretPoint;
-    return true;
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const SuggestCaretOptions& aOptions) {
-    MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
-    MOZ_ASSERT(
-        !aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt));
-    if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
-        !mCaretPoint.IsSet()) {
-      return false;
-    }
-    aPointToPutCaret = UnwrapCaretPoint();
-    return true;
-  }
-  bool CopyCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const HTMLEditor& aHTMLEditor,
-                        const SuggestCaretOptions& aOptions) const;
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const HTMLEditor& aHTMLEditor,
-                        const SuggestCaretOptions& aOptions);
-
   SplitNodeResult() = delete;
   SplitNodeResult(const SplitNodeResult&) = delete;
   SplitNodeResult& operator=(const SplitNodeResult&) = delete;
@@ -446,14 +363,12 @@
   SplitNodeResult(SplitNodeResult&& aSplitResult,
                   const EditorDOMPoint& aNewCaretPoint)
       : SplitNodeResult(std::move(aSplitResult)) {
-    mCaretPoint = aNewCaretPoint;
-    mHandledCaretPoint = false;
+    SetCaretPoint(aNewCaretPoint);
   }
   SplitNodeResult(SplitNodeResult&& aSplitResult,
                   EditorDOMPoint&& aNewCaretPoint)
       : SplitNodeResult(std::move(aSplitResult)) {
-    mCaretPoint = std::move(aNewCaretPoint);
-    mHandledCaretPoint = false;
+    SetCaretPoint(std::move(aNewCaretPoint));
   }
 
   /**
@@ -472,35 +387,13 @@
   SplitNodeResult(nsIContent& aNewNode, nsIContent& aSplitNode,
                   SplitNodeDirection aDirection,
                   const Maybe<EditorDOMPoint>& aNewCaretPoint = Nothing())
-      : mPreviousNode(aDirection == SplitNodeDirection::LeftNodeIsNewOne
-                          ? &aNewNode
-                          : &aSplitNode),
-        mNextNode(aDirection == SplitNodeDirection::LeftNodeIsNewOne
-                      ? &aSplitNode
-                      : &aNewNode),
-        mCaretPoint(aNewCaretPoint.isSome()
-                        ? aNewCaretPoint.ref()
-                        : EditorDOMPoint::AtEndOf(mPreviousNode)),
+      : CaretPoint(aNewCaretPoint.isSome()
+                       ? aNewCaretPoint.ref()
+                       : EditorDOMPoint::AtEndOf(*PreviousNode(
+                             aDirection, &aNewNode, &aSplitNode))),
+        mPreviousNode(PreviousNode(aDirection, &aNewNode, &aSplitNode)),
+        mNextNode(NextNode(aDirection, &aNewNode, &aSplitNode)),
         mDirection(aDirection) {}
-  SplitNodeResult(nsCOMPtr<nsIContent>&& aNewNode,
-                  nsCOMPtr<nsIContent>&& aSplitNode,
-                  SplitNodeDirection aDirection,
-                  const Maybe<EditorDOMPoint>& aNewCaretPoint = Nothing())
-      : mPreviousNode(aDirection == SplitNodeDirection::LeftNodeIsNewOne
-                          ? std::move(aNewNode)
-                          : std::move(aSplitNode)),
-        mNextNode(aDirection == SplitNodeDirection::LeftNodeIsNewOne
-                      ? std::move(aSplitNode)
-                      : std::move(aNewNode)),
-        mCaretPoint(aNewCaretPoint.isSome()
-                        ? aNewCaretPoint.ref()
-                        : (MOZ_LIKELY(mPreviousNode)
-                               ? EditorDOMPoint::AtEndOf(mPreviousNode)
-                               : EditorDOMPoint())),
-        mDirection(aDirection) {
-    MOZ_DIAGNOSTIC_ASSERT(mPreviousNode);
-    MOZ_DIAGNOSTIC_ASSERT(mNextNode);
-  }
 
   SplitNodeResult ToHandledResult() const {
     mHandledCaretPoint = true;
@@ -531,8 +424,8 @@
     result.mPreviousNode = &aNotSplitNode;
     // Caret should be put at the last split point instead of current node.
     if (aDeeperSplitNodeResult) {
-      result.mCaretPoint = aDeeperSplitNodeResult->mCaretPoint;
-      aDeeperSplitNodeResult->mHandledCaretPoint = true;
+      result.SetCaretPoint(aDeeperSplitNodeResult->CaretPointRef());
+      aDeeperSplitNodeResult->IgnoreCaretPointSuggestion();
     }
     return result;
   }
@@ -544,8 +437,8 @@
     result.mNextNode = &aNotSplitNode;
     // Caret should be put at the last split point instead of current node.
     if (aDeeperSplitNodeResult) {
-      result.mCaretPoint = aDeeperSplitNodeResult->mCaretPoint;
-      aDeeperSplitNodeResult->mHandledCaretPoint = true;
+      result.SetCaretPoint(aDeeperSplitNodeResult->CaretPointRef());
+      aDeeperSplitNodeResult->IgnoreCaretPointSuggestion();
     }
     return result;
   }
@@ -559,8 +452,8 @@
     result.mGivenSplitPoint = aGivenSplitPoint;
     // Caret should be put at the last split point instead of current node.
     if (aDeeperSplitNodeResult) {
-      result.mCaretPoint = aDeeperSplitNodeResult->mCaretPoint;
-      aDeeperSplitNodeResult->mHandledCaretPoint = true;
+      result.SetCaretPoint(aDeeperSplitNodeResult->CaretPointRef());
+      aDeeperSplitNodeResult->IgnoreCaretPointSuggestion();
     }
     return result;
   }
@@ -575,25 +468,40 @@
   static inline SplitNodeResult MergeWithDeeperSplitNodeResult(
       SplitNodeResult&& aSplitNodeResult,
       const SplitNodeResult& aDeeperSplitNodeResult) {
-    aSplitNodeResult.mHandledCaretPoint = false;
-    aDeeperSplitNodeResult.mHandledCaretPoint = true;
+    aSplitNodeResult.UnmarkAsHandledCaretPoint();
+    aDeeperSplitNodeResult.IgnoreCaretPointSuggestion();
     if (aSplitNodeResult.DidSplit() ||
         !aDeeperSplitNodeResult.mCaretPoint.IsSet()) {
       return std::move(aSplitNodeResult);
     }
     SplitNodeResult result(std::move(aSplitNodeResult));
-    result.mCaretPoint = aDeeperSplitNodeResult.mCaretPoint;
+    result.SetCaretPoint(aDeeperSplitNodeResult.CaretPointRef());
     return result;
   }
 
 #ifdef DEBUG
-  ~SplitNodeResult() { MOZ_ASSERT(!mCaretPoint.IsSet() || mHandledCaretPoint); }
+  ~SplitNodeResult() {
+    MOZ_ASSERT(!HasCaretPointSuggestion() || CaretPointHandled());
+  }
 #endif
 
  private:
   explicit SplitNodeResult(SplitNodeDirection aDirection)
       : mDirection(aDirection) {}
 
+  // Helper methods to consider previous/next node from new/old node and split
+  // direction.
+  static nsIContent* PreviousNode(SplitNodeDirection aDirection,
+                                  nsIContent* aNewOne, nsIContent* aOldOne) {
+    return aDirection == SplitNodeDirection::LeftNodeIsNewOne ? aNewOne
+                                                              : aOldOne;
+  }
+  static nsIContent* NextNode(SplitNodeDirection aDirection,
+                              nsIContent* aNewOne, nsIContent* aOldOne) {
+    return aDirection == SplitNodeDirection::LeftNodeIsNewOne ? aOldOne
+                                                              : aNewOne;
+  }
+
   // When methods which return this class split some nodes actually, they
   // need to set a set of left node and right node to this class.  However,
   // one or both of them may be moved or removed by mutation observer.
@@ -677,7 +585,7 @@
  * SplitRangeOffFromNodeResult class is a simple class for methods which split a
  * node at 2 points for making part of the node split off from the node.
  *****************************************************************************/
-class MOZ_STACK_CLASS SplitRangeOffFromNodeResult final {
+class MOZ_STACK_CLASS SplitRangeOffFromNodeResult final : public CaretPoint {
  public:
   /**
    * GetLeftContent() returns new created node before the part of quarried out.
@@ -752,45 +660,13 @@
     return ContentNodeType::FromNodeOrNull(GetRightmostContent());
   }
 
-  /**
-   * Suggest caret position to aHTMLEditor.
-   */
-  [[nodiscard]] MOZ_CAN_RUN_SCRIPT nsresult SuggestCaretPointTo(
-      const HTMLEditor& aHTMLEditor, const SuggestCaretOptions& aOptions) const;
-
-  /**
-   * IgnoreCaretPointSuggestion() should be called if the method does not want
-   * to use caret position recommended by this instance.
-   */
-  void IgnoreCaretPointSuggestion() const { mHandledCaretPoint = true; }
-
-  bool HasCaretPointSuggestion() const { return mCaretPoint.IsSet(); }
-  constexpr EditorDOMPoint&& UnwrapCaretPoint() {
-    mHandledCaretPoint = true;
-    return std::move(mCaretPoint);
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const SuggestCaretOptions& aOptions) {
-    MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
-    MOZ_ASSERT(
-        !aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt));
-    if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
-        !mCaretPoint.IsSet()) {
-      return false;
-    }
-    aPointToPutCaret = UnwrapCaretPoint();
-    return true;
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const HTMLEditor& aHTMLEditor,
-                        const SuggestCaretOptions& aOptions);
-
   SplitRangeOffFromNodeResult() = delete;
 
   SplitRangeOffFromNodeResult(nsIContent* aLeftContent,
                               nsIContent* aMiddleContent,
                               nsIContent* aRightContent)
-      : mLeftContent(aLeftContent),
+      : CaretPoint(),
+        mLeftContent(aLeftContent),
         mMiddleContent(aMiddleContent),
         mRightContent(aRightContent) {}
 
@@ -798,20 +674,20 @@
                               nsIContent* aMiddleContent,
                               nsIContent* aRightContent,
                               EditorDOMPoint&& aPointToPutCaret)
-      : mLeftContent(aLeftContent),
+      : CaretPoint(std::move(aPointToPutCaret)),
+        mLeftContent(aLeftContent),
         mMiddleContent(aMiddleContent),
-        mRightContent(aRightContent),
-        mCaretPoint(std::move(aPointToPutCaret)) {}
+        mRightContent(aRightContent) {}
 
   SplitRangeOffFromNodeResult(const SplitRangeOffFromNodeResult& aOther) =
       delete;
   SplitRangeOffFromNodeResult& operator=(
       const SplitRangeOffFromNodeResult& aOther) = delete;
   SplitRangeOffFromNodeResult(SplitRangeOffFromNodeResult&& aOther) noexcept
-      : mLeftContent(std::move(aOther.mLeftContent)),
+      : CaretPoint(aOther.UnwrapCaretPoint()),
+        mLeftContent(std::move(aOther.mLeftContent)),
         mMiddleContent(std::move(aOther.mMiddleContent)),
-        mRightContent(std::move(aOther.mRightContent)),
-        mCaretPoint(std::move(aOther.mCaretPoint)) {
+        mRightContent(std::move(aOther.mRightContent)) {
     MOZ_ASSERT(!aOther.mMovedContent);
   }
   SplitRangeOffFromNodeResult& operator=(SplitRangeOffFromNodeResult&& aOther) =
@@ -819,7 +695,7 @@
 
 #ifdef DEBUG
   ~SplitRangeOffFromNodeResult() {
-    MOZ_ASSERT(!mCaretPoint.IsSet() || mHandledCaretPoint);
+    MOZ_ASSERT(!HasCaretPointSuggestion() || CaretPointHandled());
   }
 #endif
 
@@ -828,11 +704,6 @@
   MOZ_KNOWN_LIVE nsCOMPtr<nsIContent> mMiddleContent;
   MOZ_KNOWN_LIVE nsCOMPtr<nsIContent> mRightContent;
 
-  // The point which is a good point to put caret from point of view the
-  // splitter.
-  EditorDOMPoint mCaretPoint;
-
-  bool mutable mHandledCaretPoint = false;
   bool mutable mMovedContent = false;
 };
 
@@ -840,7 +711,7 @@
  * SplitRangeOffResult class is a simple class for methods which splits
  * specific ancestor elements at 2 DOM points.
  *****************************************************************************/
-class MOZ_STACK_CLASS SplitRangeOffResult final {
+class MOZ_STACK_CLASS SplitRangeOffResult final : public CaretPoint {
  public:
   constexpr bool Handled() const { return mHandled; }
 
@@ -851,39 +722,6 @@
    */
   constexpr const EditorDOMRange& RangeRef() const { return mRange; }
 
-  /**
-   * Suggest caret position to aHTMLEditor.
-   */
-  [[nodiscard]] MOZ_CAN_RUN_SCRIPT nsresult SuggestCaretPointTo(
-      const HTMLEditor& aHTMLEditor, const SuggestCaretOptions& aOptions) const;
-
-  /**
-   * IgnoreCaretPointSuggestion() should be called if the method does not want
-   * to use caret position recommended by this instance.
-   */
-  void IgnoreCaretPointSuggestion() const { mHandledCaretPoint = true; }
-
-  bool HasCaretPointSuggestion() const { return mCaretPoint.IsSet(); }
-  constexpr EditorDOMPoint&& UnwrapCaretPoint() {
-    mHandledCaretPoint = true;
-    return std::move(mCaretPoint);
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const SuggestCaretOptions& aOptions) {
-    MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
-    MOZ_ASSERT(
-        !aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt));
-    if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
-        !mCaretPoint.IsSet()) {
-      return false;
-    }
-    aPointToPutCaret = UnwrapCaretPoint();
-    return true;
-  }
-  bool MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
-                        const HTMLEditor& aHTMLEditor,
-                        const SuggestCaretOptions& aOptions);
-
   SplitRangeOffResult() = delete;
 
   /**
@@ -906,7 +744,8 @@
   SplitRangeOffResult(EditorDOMRange&& aTrackedRange,
                       SplitNodeResult&& aSplitNodeResultAtStart,
                       SplitNodeResult&& aSplitNodeResultAtEnd)
-      : mRange(std::move(aTrackedRange)),
+      : CaretPoint(),
+        mRange(std::move(aTrackedRange)),
         mHandled(aSplitNodeResultAtStart.Handled() ||
                  aSplitNodeResultAtEnd.Handled()) {
     MOZ_ASSERT(mRange.StartRef().IsSet());
@@ -914,12 +753,14 @@
     // The given results are created for creating this instance so that the
     // caller may not need to handle with them.  For making who taking the
     // responsible clearer, we should move them into this constructor.
+    EditorDOMPoint pointToPutCaret;
     SplitNodeResult splitNodeResultAtStart(std::move(aSplitNodeResultAtStart));
     SplitNodeResult splitNodeResultAtEnd(std::move(aSplitNodeResultAtEnd));
     splitNodeResultAtStart.MoveCaretPointTo(
-        mCaretPoint, {SuggestCaret::OnlyIfHasSuggestion});
-    splitNodeResultAtEnd.MoveCaretPointTo(mCaretPoint,
+        pointToPutCaret, {SuggestCaret::OnlyIfHasSuggestion});
+    splitNodeResultAtEnd.MoveCaretPointTo(pointToPutCaret,
                                           {SuggestCaret::OnlyIfHasSuggestion});
+    SetCaretPoint(std::move(pointToPutCaret));
   }
 
   SplitRangeOffResult(const SplitRangeOffResult& aOther) = delete;
@@ -936,12 +777,7 @@
   // the node might have gone with another DOM tree mutation.  So, be careful
   // if you do it.
 
-  // The point which is a good point to put caret from point of view the
-  // splitter.
-  EditorDOMPoint mCaretPoint;
-
   bool mHandled;
-  bool mutable mHandledCaretPoint = false;
 };
 
 /******************************************************************************